src/tasks.ts

```diff
@@ -0,0 +1,329 @@
+        }
+    }
+
+    debug('Loading validators');
+    await loadValidators(Number(targetEpoch));
+
+    debug('Fetching duties');
+    // TODO: fix range
+    for (const epoch of R.range(Number(sourceEpoch), Number(targetEpoch))) {
+        await lookupCommittees(epoch);
+    }
+
+    debug('Check duties');
+    // TODO: fix range
+    for (const epoch of R.range(Number(sourceEpoch), Number(targetEpoch))) {
+        await checkEpochSlots(epoch);
+    }
+
+    // TODO: read the threshold from somewhere
+    Cache.excludeFromStats(belowThreshold(0.9));
+
+    const feeShares = await Shared.STETH.sharesOf(Shared.CONFIG.CSM_ADDRESS);
+    const distributed = distributeFees(feeShares + 42n);
+
+    const leafs = R.sortBy(R.prop(0), [...Cache.rewards.entries()]);
+    debug('Rewards distribution', leafs);
+    const tree = buildTree(leafs);
+    debug('Report tree root', tree.root);
+    saveArtifact(Number(sourceEpoch), Number(targetEpoch), tree);
+
+    // TODO: Simulate report sending
+    await Shared.ORACLE.connect(Shared.SIGNER).submitReport.staticCall(
+        targetEpoch,
+        tree.root,
+        distributed,
+        '', // CID TBA, but can be precalculated
+    );
+
+    const cid = await uploadTree(tree);
+    // TODO: send a report
+    await Shared.ORACLE.connect(Shared.SIGNER).submitReport(
+        targetEpoch,
+        tree.root,
+        distributed,
+        cid.toString(),
+    );
+
+    debug('Report sent, storing the artifact');
+    saveArtifact(Number(sourceEpoch), Number(targetEpoch), tree);
+}
+
+async function loadValidators(epoch: Epoch) {
+    const vals = await getModuleValidators(epochLastSlot(epoch));
+    debug('Found validators:', vals.length);
+    for (const v of vals) {
+        Cache.indexToPubkey.set(v.index, toHex(v.validator.pubkey));
+    }
+}
+
+async function lookupCommittees(epoch: Epoch) {
+    debug('Looking up committees for epoch', epoch);
+    const epochCommitees = await getEpochCommmittees(epoch);
+    epochCommitees.map((committee) => {
+        committee.validators.forEach((index, pos) => {
+            if (Cache.isTrackedValidator(index)) {
+                Cache.addValidatorToSlotDuty(index, committee.slot, pos);
+            }
+        });
+    });
+}
+
+async function checkEpochSlots(epoch: Epoch) {
+    debug('Checking epoch slots', epoch);
+    for (const slot of iterSlotsInEpoch(epoch)) {
+        const atts = await getSlotAttestations(slot);
+        atts.forEach((a) => {
+            const d = Cache.duties.get(a.data.slot);
+            if (d) {
+                d.forEach((c, i) => {
+                    if (a.aggregationBits.get(c.position)) {
+                        const p = Cache.stats.get(
+                            Cache.indexToPubkey.get(c.validatorIndex),
+                        );
+                        if (p) {
+                            p.missedAttestations--;
+                        }
+                        delete d[i];
```

Comment:
Doesn't it break the loop? you removing an element from the current iterable

Reply:
Here I've removed the deletion. We will see if it cause any performance issues later.

Reply:
It can be strange a little bit a validator includes 1 attestation multiple time. I've to think about it.